memflow/src/plugins/mod.rs

```diff
@@ -6,35 +6,160 @@ This module contains functions to interface with dynamically loaded connectors a
 pub mod args;
 #[doc(hidden)]
 pub use args::Args;
 
+// cbindgen fails to properly parse this as return type
+pub type OptionVoid = Option<&'static mut std::ffi::c_void>;
+
 pub mod connector;
 pub use connector::{
-    ConnectorBaseTable, ConnectorDescriptor, ConnectorFunctionTable, ConnectorInstance,
-    PhysicalMemoryFunctionTable, MEMFLOW_CONNECTOR_VERSION,
+    ConnectorDescriptor, ConnectorFunctionTable, ConnectorInstance, LoadableConnector,
+    OpaquePhysicalMemoryFunctionTable,
+};
+pub type ConnectorInputArg = <LoadableConnector as Loadable>::InputArg;
+
+pub mod os;
+pub use os::{LoadableOS, OSInstance, OSLayerDescriptor, OpaqueOSFunctionTable};
+pub type OSInputArg = <LoadableOS as Loadable>::InputArg;
+
+pub(crate) mod util;
+pub use util::create_bare;
+
+pub mod virt_mem;
+pub use virt_mem::{
+    OpaqueVirtualMemoryFunctionTable, VirtualMemoryFunctionTable, VirtualMemoryInstance,
 };
+pub(crate) mod arc;
+pub(crate) use arc::{CArc, COptArc};
 
-use crate::error::*;
+use crate::error::{Result, *};
 
 use log::*;
+use std::ffi::c_void;
 use std::fs::read_dir;
 use std::path::{Path, PathBuf};
-use std::sync::Arc;
 
 use libloading::Library;
 
+/// Exported memflow plugins version
+pub const MEMFLOW_PLUGIN_VERSION: i32 = 8;
+
+/// Utility typedef for better cbindgen
+///
+/// TODO: remove when fixed in cbindgen
+pub type OptionMut<T> = Option<&'static mut T>;
+
+/// Opaque version of `GenericBaseTable` for FFI purposes
+pub type OpaqueBaseTable = GenericBaseTable<c_void>;
+/// Opaque version of `GenericCloneTable` for FFI purposes
+pub type OpaqueCloneTable = GenericCloneTable<c_void>;
+
+impl Copy for OpaqueCloneTable {}
+
+impl Clone for OpaqueCloneTable {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+/// Generic function for cloning past FFI boundary
+#[repr(C)]
+pub struct GenericCloneTable<T: 'static> {
+    pub clone: extern "C" fn(this: &T) -> OptionMut<T>,
+}
+
+impl<T: Clone> Default for GenericCloneTable<T> {
+    fn default() -> Self {
+        Self {
+            clone: util::c_clone::<T>,
+        }
+    }
+}
+
+impl<T: Clone> GenericCloneTable<T> {
+    pub fn into_opaque(self) -> OpaqueCloneTable {
+        unsafe { std::mem::transmute(self) }
+    }
+}
+
+impl Copy for OpaqueBaseTable {}
+
+impl Clone for OpaqueBaseTable {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+/// Base table for most objects that are cloneable and droppable.
+#[repr(C)]
+pub struct GenericBaseTable<T: 'static> {
+    clone: GenericCloneTable<T>,
+    pub drop: unsafe extern "C" fn(this: &mut T),
+}
+
+impl<T: Clone> Default for GenericBaseTable<T> {
+    fn default() -> Self {
+        Self {
+            clone: GenericCloneTable::<T>::default(),
+            drop: util::c_drop::<T>,
+        }
+    }
+}
+
+impl<T: Clone> GenericBaseTable<T> {
+    pub fn into_opaque(self) -> OpaqueBaseTable {
+        unsafe { std::mem::transmute(self) }
+    }
+}
+
+/// Describes a FFI safe option
+#[repr(C)]
+pub enum COption<T> {
```

Comment:
Apart from COption and the other types it would also be good to move the VTable functionality into this separate crate. A complex proc macro could be leveraged to generate all the table boilerplate code for added convenience.